torch/csrc/jit/ir/alias_analysis.cpp

```diff
@@ -390,6 +392,69 @@ std::string AliasDb::toString() const {
   return ss.str();
 }
 
+bool AliasDb::dumpToDot(const char* filename) const {
+  std::ofstream dot(filename);
+  if (!dot.good()) {
+    std::cout << "Failed to open the IR graph file: '" << filename << "'\n";
```

Comment:
I think we typically use const std::string& and use exceptions for signaling errors rather than return values.

I think an even better (and more consistent with the rest of the JIT API) way would be to just define the op << .
Then the string dump (great idea) you mentioned on slack could use that in the string stream and opening the file yourself when dumping doesn't seem too much of a hassle (and then people can also dump to stdout or over the network or whatever if they want to).

Reply:
In general I agree. One can also point out that printing an message to console _and_ returning an error is bad form.

In this case, throwing an exception doesn't play nicely in a debugger, which is one of the intended use-cases. Normally in the debugger a simple message would do, but the return bool is a compromise to allow some basic composition (so the caller can still tell if it faild)

Same thing with the operator<< overloading - doesn't work well in debuggers.

Reply:
Yeah, good points. Happily I don't get to decide on it, anyways. But the string would be possible?

Reply:
> But the string would be possible?

Using `std::string` for the `filename` parameter? Same story: from the C++ side, it seems a minor change, but for a debugger, evaluating `foo("bar")` when foo takes a `const std::string&` is far from trivial and not all the debugger support it (my workflow involves using gdb, lldb, windbg/cdb & visual studio regularly and I've learned the hard way to keep debugging helpers as simple as possible).